### PR TITLE
[core] add reloadAppAsync support

### PR DESCRIPTION
**Pull request by Kudo**
# Why

close ENG-11482

# How

introduce `reloadAppAsync()` from the `expo` package

some reflection code could be removed after https://github.com/facebook/react-native/pull/44223 landed

# Test Plan

add NCL `ExpoCoreModule` to test the reload function

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]

**Review by ide — COMMENTED** [review]
If the app has installed a new update, will this new `reloadAppAsync()` method apply it? I anticipate two questions people will have are: why are there two reload methods and which one should I use?

**Comment by Kudo** [comment]
> If the app has installed a new update, will this new `reloadAppAsync()` method apply it? I anticipate two questions people will have are: why are there two reload methods and which one should I use?

no. `reloadAppAsync()` is just to reload the app. `Updates.reloadAsync()` will further change the loading bundle if there're newer updates. the use case is slightly different. since it's `Updates.reloadAsync()`, it has updates concept involved. does the name clear enough?

**Comment by ide** [comment]
I think the names are clear but also think it would be helpful in the docs for each function to explain how it is different than the other function.

**Review by tsapeta — APPROVED** [review]

**Comment by Kudo** [comment]
> I think the names are clear but also think it would be helpful in the docs for each function to explain how it is different than the other function.

tried to update here. please help to review if the description is clear
- https://github.com/expo/expo/blob/654d43c5adaf223a9976a41a9f99de86cc27d96b/packages/expo-modules-core/src/reload.ts#L6-L7
- https://github.com/expo/expo/blob/654d43c5adaf223a9976a41a9f99de86cc27d96b/packages/expo-updates/src/Updates.ts#L142-L143

File: packages/expo-updates/src/Updates.ts [hunk 1]
```diff
@@ -139,6 +139,8 @@ const manualUpdatesInstructions =
  * Instructs the app to reload using the most recently downloaded version. This is useful for
  * triggering a newly downloaded update to launch without the user needing to manually restart the
  * app.
+ * Unlike `Expo.reloadAppAsync()` serving from the `expo` package,
+ * this function not only reloads the app but also changes the loading JavaScript bundle to the most recently downloaded one.
```
Comment (ide) [inline packages/expo-updates/src/Updates.ts:143]:
```suggestion
 * Unlike `Expo.reloadAppAsync()` provided by the `expo` package,
 * this function not only reloads the app but also changes the loaded JavaScript bundle to that of the most recently downloaded update.
```

File: packages/expo-modules-core/src/reload.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Reloads the app.
+ *
+ * This function should work for both release and debug builds.
+ *
+ * Unlike the [`Updates.reloadAsync()`](https://docs.expo.dev/versions/latest/sdk/updates/#updatesreloadasync),
+ * this function does not change the loading bundle. It only reloads with the same JavaScript bundle.
```
Comment (ide) [inline packages/expo-modules-core/src/reload.ts:7]:
```suggestion
 * This function works for both release and debug builds.
 *
 * Unlike [`Updates.reloadAsync()`](https://docs.expo.dev/versions/latest/sdk/updates/#updatesreloadasync),
 * this function does not use a new update even if one is available. It only reloads the app using the same JavaScript bundle that is currently running.
```

**Comment by Kudo** [comment]
i was waiting for 0.74.1 for the changes from core and not using the reflection on android. since we may like to submit expo-go beforehand, i'm going to merge this first. later when we upgrade 0.74.1, i will remove the reflection.

**Comment by laurienicholas** [comment]
After a recent upgrade my iOS compilation is currently failing with this error:

```
❌  (/[@@@]/node_modules/expo-modules-core/ios/Core/Modules/CoreModule.swift:44:7)

  42 |
  43 |     AsyncFunction("reloadAppAsync") { (reason: String) in
> 44 |       RCTTriggerReloadCommandListeners(reason)
     |       ^ cannot find 'RCTTriggerReloadCommandListeners' in scope
  45 |     }
  46 |   }
  47 | }
```

I believe it may be related to this PR? Could anyone with more knowledge of the inner workings provide some guidance?

**Comment by brentvatne** [comment]
@laurienicholas - run `npx expo-doctor@latest` to verify your dependencies are all aligned. if that doesn't answer your question, please create a new issue with a minimal reproducible example.

**Pull request merged**